framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,670 @@
+    }
+
+    /** Return true if this range contains no values. */
+    public boolean isNothing() {
+        return this == NOTHING;
+    }
+
+    /**
+     * Creates a 32-bit integral range from this range. Returns INT_EVERYTHING if this range doesn't
+     * fit in 32-bit values.
+     */
+    public Range intRange() {
+        if (from < Integer.MIN_VALUE || to > Integer.MAX_VALUE) {
+            return INT_EVERYTHING;
+        } else {
+            return this;
+        }
+    }
+
+    /**
+     * Creates a 16-bit short range from this range. Returns SHORT_EVERYTHING if this range doesn't
+     * fit in 16-bit values.
+     */
+    public Range shortRange() {
+        if (from < Short.MIN_VALUE || to > Short.MAX_VALUE) {
+            return INT_EVERYTHING;
+        } else {
+            return this;
+        }
+    }
+
+    /** Returns true if the element is contained in this range. */
+    public boolean contains(long element) {
+        return from <= element && element <= to;
+    }
+
+    /**
+     * Returns the smallest range that includes all values contained in either of the two ranges. We
+     * call this the union of two ranges.
+     *
+     * @param right a range to union with this range
+     * @return a range resulting from the union of the specified range and this range
+     */
+    public Range union(Range right) {
+        if (this.isNothing()) {
+            return right;
+        } else if (right.isNothing()) {
+            return this;
+        }
+
+        long resultFrom = Math.min(from, right.from);
+        long resultTo = Math.max(to, right.to);
+        return new Range(resultFrom, resultTo);
+    }
+
+    /**
+     * Returns the smallest range that includes all values contained in both of the two ranges. We
+     * call this the intersection of two ranges. If there is no overlap between the two ranges,
+     * returns an empty range.
+     *
+     * @param right the range to intersect with this range
+     * @return a range resulting from the intersection of the specified range and this range
+     */
+    public Range intersect(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        long resultFrom = Math.max(from, right.from);
+        long resultTo = Math.min(to, right.to);
+        return createRangeOrNothing(resultFrom, resultTo);
+    }
+
+    /**
+     * Returns the smallest range that includes all possible values resulting from adding an
+     * arbitrary value in the specified range to an arbitrary value in this range. We call this We
+     * call this the addition of two ranges.
+     *
+     * @param right a range to be added to this range
+     * @return the range resulting from the addition of the specified range and this range
+     */
+    public Range plus(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (this.isWithinInteger() && right.isWithinInteger()) {
```

Comment:
The `plus` and `minus` routines can use integers for more values.  In particular, adding or removing can only add one more bit to the length of the binary representation.  The current implementation is correct, just not as efficient as it could be.  You can leave it if you want to.